core/silkworm/types/account.cpp

```diff
@@ -84,85 +84,109 @@ size_t Account::encoding_length_for_storage() const {
-    uint8_t field_set = encoded[0];
-    size_t pos{1};
-
-    if (field_set & 1) {
-        pos += encoded[pos++];
+static inline std::pair<uint8_t, rlp::DecodingResult> validate_encoded_head(ByteView& encoded_payload) noexcept {
+    if (encoded_payload.empty()) {
+        return {0, rlp::DecodingResult::kOk};
     }
-    if (field_set & 2) {
-        pos += encoded[pos++];
+    if (encoded_payload[0] && encoded_payload.length() == 1) {
+        // Must be at least 2 bytes : field_set + len of payload
+        return {encoded_payload[0], rlp::DecodingResult::kInputTooShort};
     }
-    if (field_set & 4) {
-        // Incarnation has been found.
-        uint8_t len = encoded[pos++];
-        const std::optional<uint64_t> incarnation{endian::from_big_compact<uint64_t>(encoded.substr(pos, len))};
-        if (incarnation == std::nullopt) {
-            return {0, rlp::DecodingResult::kOverflow};
-        }
-        return {*incarnation, rlp::DecodingResult::kOk};
+    if (encoded_payload[0] > 15) {
+        // Can only be at max 1 | 2 | 4 | 8
+        return {encoded_payload[0], rlp::DecodingResult::kInvalidFieldset};
     }
-    return {0, rlp::DecodingResult::kOk};
+
+    return {encoded_payload[0], rlp::DecodingResult::kOk};
 }
 
-std::pair<Account, rlp::DecodingResult> decode_account_from_storage(ByteView encoded) noexcept {
+std::pair<Account, rlp::DecodingResult> Account::from_encoded_storage(ByteView encoded_payload) noexcept {
     Account a{};
-    if (encoded.empty()) {
+    const auto [field_set, err] = validate_encoded_head(encoded_payload);
+    if (err != rlp::DecodingResult::kOk) {
+        return {a, err};
+    } else if (!field_set) {
         return {a, rlp::DecodingResult::kOk};
     }
 
-    uint8_t field_set = encoded[0];
     size_t pos{1};
-
-    if (field_set & 1) {
-        uint8_t len = encoded[pos++];
-        if (encoded.length() < pos + len) {
-            return {a, rlp::DecodingResult::kInputTooShort};
-        }
-        const std::optional<uint64_t> nonce{endian::from_big_compact<uint64_t>(encoded.substr(pos, len))};
-        if (nonce == std::nullopt) {
-            return {a, rlp::DecodingResult::kOverflow};
+    for (int i{1}; i < 16; i *= 2) {
+        if (field_set & i) {
+            uint8_t len = encoded_payload[pos++];
+            if (encoded_payload.length() < pos + len) {
+                return {a, rlp::DecodingResult::kInputTooShort};
+            }
+            switch (i) {
+                case 1: {
+                    const std::optional<uint64_t> nonce{
+                        endian::from_big_compact<uint64_t>(encoded_payload.substr(pos, len))};
+                    if (nonce == std::nullopt) {
+                        return {a, rlp::DecodingResult::kLeadingZero};
+                    }
+                    a.nonce = *nonce;
+                } break;
+                case 2:
+                    std::memcpy(&as_bytes(a.balance)[32 - len], &encoded_payload[pos], len);
```

Comment:
Let's use `endian::from_big_compact<intx::uint256>` for consistency

Reply:
Sure ... I have question though.
Actually `from_big_compact` returns std::nullopt in two cases:

- either size of target is < size of data string (overflow)
- or there are leading zeroes and are not allowed

How we could efficiently make a distinction ?
Wouldn't it be better to transform `from_big_compact` into a parser into the rlp namespace ?
E.g.

```
template <typename UnsignedInteger>
static rlp::DecodingResult parse_big_compact(ByteView data, UnsignedInteger& target, bool allow_leading_zeros = false) {
 [...]
 target = be:load(x)
 return rlp::DecodingResult::kOk
)
```

Reply:
Perhaps. I don't have a strong view on that. An argument against moving `parse_big_compact` into RLP is that it's not particularly RLP-specific, but rather a useful low-level function. If you decide to move it into RLP, please move it alongside `to_big_compact` and in a separate PR.